jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -355,51 +357,72 @@ void checkAccess(VarHandle handle) throws IllegalAccessException {
 				throw new IllegalAccessException(this.toString());
 			}
 			
-			checkAccess(handle.getDefiningClass(), handle.getFieldName(), handle.getModifiers(), null);
+			Class<?> definingClass = handle.getDefiningClass();
+			checkClassAccess(definingClass);
+			checkMemberAccess(definingClass, handle.getFieldName(), handle.getModifiers(), null, null);
 		}
 		
 		void checkAccess(Class<?> clazz) throws IllegalAccessException {
 			if (INTERNAL_PRIVILEGED == accessMode) {
 				// Full access for use by MH implementation.
 				return;
 			}
-					
+			
 			if (NO_ACCESS == accessMode) {
 				throw new IllegalAccessException(this.toString());
 			}
 			
 			checkClassAccess(clazz);
 		}
 		/*[ENDIF]*/
-						
+		
+		/* Do the access check of class based on whether it is an nested class or not */
+		private void checkClassAccess(Class<?> targetClass) throws IllegalAccessException {
+			if (targetClass.isMemberClass()) {
+				checkMemberAccess(targetClass.getEnclosingClass(), targetClass.getName(), targetClass.getModifiers(), targetClass, null);
+			} else {
+				checkNonNestedClassAccess(targetClass);
+			}
+		}
+
 		/**
 		 * Checks whether {@link #accessClass} can access a specific member of the {@code targetClass}.
 		 * Equivalent of visible.c checkVisibility();
 		 * 
 		 * @param targetClass The {@link Class} that owns the member being accessed.
 		 * @param name The name of member being accessed.
 		 * @param memberModifiers The modifiers of the member being accessed.
+		 * @param nestedClass A nested class if exists.
 		 * @param handle A handle object (e.g. {@link MethodHandle} or {@link VarHandle}), if applicable.
 		 * 				 The object is always {@link MethodHandle} in the current implementation, so in order 
 		 * 				 to avoid extra type checking, the parameter is {@link MethodHandle}. If we need to
 		 * 				 handle {@link VarHandle} in the future, the type can be {@link Object}.
 		 * @throws IllegalAccessException If the member is not accessible.
 		 */
-		private void checkAccess(Class<?> targetClass, String name, int memberModifiers, MethodHandle handle) throws IllegalAccessException {
-			checkClassAccess(targetClass);
-
+		private void checkMemberAccess(Class<?> targetClass, String name, int memberModifiers, Class<?> nestedClass, MethodHandle handle) throws IllegalAccessException {
```

Comment:
This PR has three calls to `checkMemberAccess(targetClass, name, memberModifiers, nestedClass, handle)`:
````
checkMemberAccess(definingClass, handle.getMethodName(), handle.getModifiers(), null, handle);
checkMemberAccess(definingClass, handle.getFieldName(), handle.getModifiers(), null, null);
checkMemberAccess(targetClass.getEnclosingClass(), targetClass.getName(), targetClass.getModifiers(), targetClass, null);
````
It appears the check and handling of `nestedClass & handle` complicated `checkMemberAccess()` logic, can this method be changed to something like `checkNestClassMemberAccess()` for the case `nestedClass` not `null`, `checkMemberAccessWithMH()` for the case `handle` not `null`, and use helper methods to avoid code duplication. The idea is to simplify the checking logic to make it easier to understand what conditions are being checked for each scenario. 
@DanHeidinga pls advise if this makes sense.

P.S. pls rebase with latest master branch to get latest update in this file.

Reply:
I've played with the code and think the varhandle / methodhandle cases are similar enough that they can stay as a single `checkMemberAccess` but I think it would be clearer to split the member class version into a separate function - maybe `checkMemberClassAccess`?

In fact, when I made that change, I ended up with almost exactly the code in original `checkClassAccess` implementation with the difference being in the handling of protected interface members.

Can you comment on what the expected difference is?